### DISCUSSION TITLE
Extensible attribute syntax

### DISCUSSION THREAD

**Discussion by oowekyala** [lf-lang/lingua-franca#756] [Ideas]
I think our language would benefit from having syntax to attach metadata to all kinds of nodes. This could be used to configure code generation with much more granularity instead of only through target properties (which are global). I think it also is necessary to let every code generator define the options they need, so that the core language stays small.

For instance, for now, preambles can take a `public` or `private` modifier. This is only useful in C++, yet it reserves those keywords in all targets. Instead, with an attribute syntax (stealing that of ~Rust~ Java), we could write:
```java
@private preamble {=...=}
@public preamble {=...=}
```
in the C++ target, and purge the core language from those modifiers. The C++ code generator would be in charge of assigning meaning to those attributes.

The use cases are many:
- target-specific code gen options, that apply to only a specific syntactic element. 
   - Eg public/private C++ preambles (see above)
   - Eg attributes that determine where a preamble is to be copy-pasted. In Rust preambles are generated at the top-level of a module.     If we had an `@in_self` attribute, we could use them to define new methods on `self` instead of top-level functions.
   - Eg specify that a param is saved in a state variable (in rust this is not default) 
   ```java
   reactor Foo(@state param: u32)
   ```
   - Eg attributes on an `import` statement, to make it clear whether the preambles are imported or not (I think this is relevant in C++)
   - Eg having an attribute `@may_panic` on a reaction in Rust to instruct the runtime to recover from a panic of this reaction, or `@on_panic(ignore)` or something
- test configuration, 
  - eg writing tests for CLI parameters (#147, #659)

   ```java
   @test_args("--main-value", "2", "--main-name", "bob")
   main reactor (value: u32, name: String)
   ```
  - eg writing tests that are supposed to fail #461

   ```java
   @test_should_not_compile
   main reactor { }
   ```
- verification predicates/ postconditions/ preconditions
```java
   @invoked_exactly(2) // or invoked_at_most, etc
   reaction(something) {= ... =}
```
- configuration of connections
   ```java
   a.a @physical -> b.b;
   ```
- labeling reactions `@label("foo")` instead of `// @label foo`, that way we don't have to read comments.
- configuring compiler warnings (https://github.com/lf-lang/lingua-franca/issues/625#issuecomment-946803363)
- library evolution features, like a `@deprecated` attribute
- configuring how main reactor parameters map to CLI arguments in C++ and Rust (eg `main reactor (@cli(name="-o") output: String)`)

**Comment by edwardalee** [comment]
This is a great idea.  We could also use it in federated execution to mark reaction "unordered" without having to introduce an unordered keyword modifier for reactions.

**Comment by cmnrd** [comment]
I think the idea is great! But there is one issue with the proposed syntax. `#` denotes comments in some of our targets and hence it also denotes comments in the LF syntax. I don't think there is a good way of resolving this in xtext.

**Reply by oowekyala** [reply]
Ah yes that's right! Well I guess the java syntax `@attribute` is still free in our grammar, and is also very expressive: `@invokedExactly(2)`, `@arg(name="-o")`.
